File: lifecycledemo/src/main/java/com/example/lifecyclerdemo/BaseApp.java
package com.example.lifecyclerdemo;

import android.app.Application;
import android.util.Log;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.ProcessLifecycleOwner;

public class BaseApp extends Application {
    private static final String TAG = "App";

    @Override
    public void onCreate() {
        super.onCreate();

        ProcessLifecycleOwner.get().getLifecycle()
                .addObserver(new ApplicationLifecycleObserver());
    }

    private static class ApplicationLifecycleObserver implements LifecycleObserver {

        @OnLifecycleEvent(Lifecycle.Event.ON_START)
        private void onAppForeground() {
            Log.e(TAG, "当前在前台");
        }

        @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
        private void onAppBackground() {
            Log.e(TAG, "当前在后台");
        }
    }
}

File: lifecycledemo/src/main/java/com/example/lifecyclerdemo/BaseApp.kt
package com.example.lifecyclerdemo

import android.app.Application
import android.util.Log
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner

class BaseApp : Application() {
    override fun onCreate() {
        super.onCreate()
        ProcessLifecycleOwner.get().lifecycle
            .addObserver(ApplicationLifecycleObserver())
    }
}

class ApplicationLifecycleObserver : DefaultLifecycleObserver {

    override fun onStart(owner: LifecycleOwner) {
        super.onStart(owner)
        onAppForeground()
    }

    private fun onAppForeground() {
        Log.e("TAG", "当前在前台")
    }

    override fun onStop(owner: LifecycleOwner) {
        super.onStop(owner)
        onAppBackground()
    }

    private fun onAppBackground() {
        Log.e("TAG", "当前在后台")
    }
}